### ISSUE TITLE
Coordinates performance regression in 4.3

### ISSUE THREAD

**Issue by maxnoe** [astropy/astropy#12196]
This code runs in 0.8 seconds in astropy 4.2 but 40 seconds in astropy 4.3

```python
from ctapipe.coordinates import CameraFrame
import astropy.units as u
from astropy.coordinates import EarthLocation, AltAz, SkyCoord
from astropy.time import Time
import numpy as np
from time import perf_counter


N = 100000
location = EarthLocation.of_site('Roque de los Muchachos')
obstime = Time.now() + np.linspace(0, 1, N) * u.hour

altaz = AltAz(location=location, obstime=obstime)
alt = np.full(N, 70) * u.deg
az = np.full(N, 180) * u.deg
pointing = SkyCoord(alt=alt, az=az, frame=altaz)

x = np.random.uniform(-1, 1, N) * u.m
y = np.random.uniform(-1, 1, N) * u.m

t0 = perf_counter()
camera_frame = CameraFrame(focal_length=28 * u.m, obstime=obstime, location=location, telescope_pointing=pointing)
coords = SkyCoord(x=x, y=y, frame=camera_frame)
t1 = perf_counter()
print(t1 - t0)
```

CameraFrame implementation here: https://github.com/cta-observatory/ctapipe/blob/master/ctapipe/coordinates/camera_frame.py

I'll check and comeback with profiling

**Comment by maxnoe** [comment]
Most of the time is spend in `epv00` and `pnm06` due to a altaz to altaz, cirs to cirs transform. That apparently only happens in 4.3, not 4.2.

Why did this change?

**Comment by maxnoe** [comment]
I bisected this to the first commit of this PR here: https://github.com/astropy/astropy/pull/10994

Which is actually even twice as slow as 4.3 for this example.

**Comment by StuartLittlefair** [comment]
@maxnoe 

This isn't the easiest example for us to work with. It requires us to understand your external codebase in detail. There's no obvious coordinate transform going on here, so presumably it occurs somewhere inside the codebase for your `CameraFrame`. Is there a minimum working example you can provide which shows the slowdown?

The reason why some CIRS transforms have got much slower are explained in these comments:

https://github.com/astropy/astropy/pull/10994#issuecomment-722587326
https://github.com/astropy/astropy/pull/10994#issuecomment-722617041

The change was essential for accuracy, but only affects 2D arrays of SkyCoords for transforms that use the intermediate CIRS frame. It's not at all obvious from your examples which transforms are being triggered and why; I'd have thought you could take advantage of the direct ICRS<->AltAz route which was implemented alongside the PR you indicate and should not be slower.

**Comment by maxnoe** [comment]
@StuartLittlefair before, there was no transform at all. The only thing happening here is setting a `CoordinateAttribute` in `AltAz` frame. That triggers an `AltAz` to `AltAz` transform somehow that shouldn't be necessary.

**Comment by StuartLittlefair** [comment]
OK. I can try and look into why that's happening. Are you sure that a self-transform wasn't occurring in 4.2? Or was it just fast? My suggestion above should speed up the `AltAz` self transform but nothing I'm aware of would make a self transform necessary when it wasn't before....

**Comment by maxnoe** [comment]
Will check again and produce a more minimal example

**Comment by maxnoe** [comment]
@StuartLittlefair Yes, definitely 4.3 does a full altaz -> CIRS -> altaz transform where 4.2 does no transformation at all.

Merely creating the frame:

```python
import astropy.units as u
from astropy.coordinates import (
    EarthLocation,
    AltAz,
    SkyCoord,
    BaseCoordinateFrame,
    UnitSphericalRepresentation,
    CoordinateAttribute,
)

from astropy.time import Time
import numpy as np
from time import perf_counter

class MyFrame(BaseCoordinateFrame):
    default_representation = UnitSphericalRepresentation
    pointing = CoordinateAttribute(frame=AltAz, default=None)


N = 100000
location = EarthLocation.of_site('Roque de los Muchachos')
obstime = Time.now() + np.linspace(0, 1, N) * u.hour

altaz = AltAz(location=location, obstime=obstime)
alt = np.full(N, 70) * u.deg
az = np.full(N, 180) * u.deg
pointing = SkyCoord(alt=alt, az=az, frame=altaz)

t0 = perf_counter()
my_frame = MyFrame(pointing=pointing)
t1 = perf_counter()
print(t1 - t0)
```

With `N=100000` this takes 0.6 seconds with 4.2 and 40 seconds 4.3.

**Comment by maxnoe** [comment]
Ok, it is even more simple than that, just using a bare `.transform_to(AltAz)` is what makes the difference:

```python
import astropy.units as u
from astropy.coordinates import EarthLocation, AltAz, SkyCoord

from astropy.time import Time
import numpy as np
from time import perf_counter

N = 100000
location = EarthLocation.of_site('Roque de los Muchachos')
obstime = Time.now() + np.linspace(0, 1, N) * u.hour

altaz = AltAz(location=location, obstime=obstime)
alt = np.full(N, 70) * u.deg
az = np.full(N, 180) * u.deg
pointing = SkyCoord(alt=alt, az=az, frame=altaz)



t0 = perf_counter()
pointing.transform_to(AltAz)
t1 = perf_counter()
print(t1 - t0)
```

**Comment by maxnoe** [comment]
Transforming coordinates in a specific `AltAz(...)` frame to the generic `AltAz`should be a no-op, yet in astropy 4.3 it goes through the full chain.

**Comment by ayshih** [comment]
> Transforming coordinates in a specific `AltAz(...)` frame to the generic `AltAz`should be a no-op, yet in astropy 4.3 it goes through the full chain.

It should become a no-op after #10909 is merged, so you can disregard that particular quirk.

**Comment by ayshih** [comment]
Ah, I see that quirk is actually a major part of the problem here.  Can you change the function at the bottom of `astropy/coordinates/builtin_frames/cirs_observed_transforms.py` to this:
```python
@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, AltAz, AltAz)
@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, HADec, HADec)
def observed_to_observed(from_coo, to_frame):
    if from_coo.is_equivalent_frame(to_frame):  # loopback to the same frame
        return to_frame.realize_frame(from_coo.data)
    # for now we just implement this through CIRS to make sure we get everything
    # covered
    return from_coo.transform_to(CIRS(obstime=from_coo.obstime)).transform_to(to_frame)
```

**Comment by maxnoe** [comment]
Could that part be backported to 4.3?

**Comment by StuartLittlefair** [comment]
@ayshih does the AltAz self transform in #10909 still go through CIRS? As the comments linked above explain, this can be really very slow indeed for large arrays if skycoord. 

If so, maybe #12203 could try changing the self transform to go through ICRS instead since the direct AltAz - ICRS - AltAz transform is *much* faster when dealing with large arrays of SkyCoords

**Comment by maxnoe** [comment]
@ayshih Should we really do such a change in a bugfix? I'm not against it, just asking.

**Comment by maxnoe** [comment]
I just tested it, for 10000 coordinates at the same time but different locations, going via ICRS is almost twice as fast and the differences are < 0.1 µas.

So I would be inclined to include it. However, since the file is called `cirs_observed_transforms`, the function should probably be moved, right? What would be the appropriate place for it?

**Comment by StuartLittlefair** [comment]
I agree; there is an `icrs_observed_transforms.py` which would be the correct place.

**Comment by maxnoe** [comment]
Ok, I included the changes in #12203
